core/src/main/scala/kafka/log/Log.scala

```diff
@@ -311,6 +311,10 @@ class Log(@volatile var dir: File,
     if (!producerStateManager.isEmpty)
       throw new IllegalStateException("Producer state must be empty during log initialization")
     loadProducerState(logEndOffset, reloadFromCleanShutdown = hasCleanShutdownFile)
+    // Pass the set of loaded segment base offsets to diff against the set of producer state snapshot
+    // files. Any orphaned files should be cleaned up.
+    val segmentBaseOffsets = logSegments.map(_.baseOffset)
+    producerStateManager.cleanupOrphanSnapshotFiles(segmentBaseOffsets)
```

Comment:
So, this means that we will be removing the producer snapshot corresponding to log end offset. One of the potential impact is that if the broker has an immediate hard failure after this point, the recovery will have to rebuild the producer snapshot for the active segment, which can be expensive. We probably want to think if there is any other impact.

Reply:
I think `cleanupOrphanSnapshotFiles` always retains the last snapshot.

Reply:
Thanks. Sounds good. We probably should change the comment above to make it clear that the last snapshot is always preserved?